File: js-final-practice/practice1.js
/*Question 1
Write a function that accepts an array of numbers and returns the 2 largest numbers in the
array. The largest values should be returned in an array where the first element in the array is
the larger of the two elements. You may assume you will always be given an array that contains
at least 2 numbers.*/

function twoLargestNumbersSort(arr) {
    const sortedArr = arr.sort((a, b) => b-a);
    return [sortedArr[0], sortedArr[1]];
}

console.log(twoLargestNumbersSort([7, -2, 3, 4, 5, 1, 23, 31, 23])); // [31, 23]
console.log(twoLargestNumbersSort([7, -2, 3, 4, 5, 1, 23, 23])); // [23, 23]
// This is definitely the simplest and easiest way to do it, but not the most efficient because sorting the whole array is unncecessary.


function twoLargestNumbersNoSort(arr) {
    let largest = -Infinity;
    let secondLargest = -Infinity;
    for (let num of arr) {
        if (num > largest) {
            secondLargest = largest;
            largest = num;
        }
        else if (num > secondLargest) {
            secondLargest = num;
        }
    }
    return [largest, secondLargest];
}
console.log(twoLargestNumbersNoSort([7, -2, 3, 4, 5, 1, 23, 31, 23])); // [31, 23]
console.log(twoLargestNumbersNoSort([7, -2, 3, 4, 5, 1, 23, 23])); // [23, 23]

File: go-final-practice/practice1.go
/* Question 1
In Go, write a function named twoLargest that accepts a slice of integers and returns a slice
containing the two largest integers from the input slice. The returned slice should have the
larger integer as the first element and the second largest integer as the second element.
The input slice will always contain at least two integers.
If the largest and the second largest number in the input slice are the same, the output slice
should contain two copies of this number. */

//go:build practice1
//run with $ go run -tags practice1 practice1.go
package main

import (
	"fmt"
	"math"
)

func twoLargest(nums[]int) []int {
	largest := int(math.Inf(-1))
	secondLargest := int(math.Inf(-1))

	for _, num := range nums {
		if num > largest {
			secondLargest = largest
			largest = num
		} else if num > secondLargest {
			secondLargest = num
		}
	}
	return []int{largest, secondLargest}
}

func main() {
	fmt.Println(twoLargest([]int{7, 3, 7, 4, 2, 1, 5})) // {7, 7}
	fmt.Println(twoLargest([]int{7, 3, 9, 4, 2, 1, 5})) // {9, 7}
}